ci/datamgr.py

```diff
@@ -214,47 +195,34 @@ def impala_create_tables(con, env, *, executor=None):
         for path in con.hdfs.ls(os.path.join(test_data_dir, 'parquet'))
     ]
     for method, path in itertools.chain(parquet_files, avro_files):
-        logger.info(os.path.basename(path))
-        args = path, ALL_SCHEMAS.get(os.path.basename(path))
-        kwargs = dict(
+        logger.debug(os.path.basename(path))
+        yield executor.submit(
+            method,
+            path,
+            ALL_SCHEMAS.get(os.path.basename(path)),
             name=os.path.basename(path),
             database=env.test_data_db,
             persist=True,
         )
-        if executor is not None:
-            yield executor.submit(method, *args, **kwargs)
-        else:
-            yield method(*args, **kwargs)
 
 
-def impala_build_udfs():
+def impala_build_and_upload_udfs(hdfs, env, *, fs, executor):
+    logger.info("Building UDFs...")
+
     cwd = str(IBIS_HOME / 'ci' / 'udf')
     subprocess.run(["cmake", ".", "-G", "Ninja"], cwd=cwd)
     subprocess.run(["ninja"], cwd=cwd)
-
-
-def impala_upload_udfs(con, env):
     build_dir = IBIS_HOME / 'ci' / 'udf' / 'build'
     bitcode_dir = os.path.join(env.test_data_dir, 'udf')
-    if con.hdfs.exists(bitcode_dir):
-        con.hdfs.rm(bitcode_dir, recursive=True)
-    con.hdfs.put(str(build_dir), bitcode_dir, recursive=True)
 
+    hdfs.mkdir(bitcode_dir, create_parents=True)
 
-def load_impala_data(con, data_dir, env):
-    impala_upload_ibis_test_data_to_hdfs(con, data_dir, env)
-    impala_create_test_database(con, env)
-    with concurrent.futures.ThreadPoolExecutor(
-        max_workers=CPU_COUNT
-    ) as executor:
-        table_futures = (
-            executor.submit(table_future.result().compute_stats)
```

Comment:
Was this the main culprit?

Reply:
No, the main culprit was the fact that recursive copy of a directory from the client to HDFS was happening in series, rather than creating a task per file and allowing concurrent IO.

Reply:
`fsspec` supports async file system operations, but the `hdfs` filesystem doesn't implement it.